src/backends/policy_engine/souffle/datalog_lowering_visitor_test.cc

```diff
@@ -0,0 +1,243 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#include "datalog_lowering_visitor.h"
+
+#include "absl/strings/string_view.h"
+#include "src/common/testing/gtest.h"
+#include "src/ir/block_builder.h"
+#include "src/ir/datalog/value.h"
+#include "src/ir/module.h"
+
+namespace raksha::backends::policy_engine::souffle {
+
+using testing::Combine;
+using testing::TestWithParam;
+using testing::ValuesIn;
+
+struct IrAndDatalogOperationPairs {
+  const ir::Operation *ir_operation;
+  const ir::datalog::IsOperationFact *datalog_is_operation;
+};
+
+class IrOperationLoweringTest
+    : public TestWithParam<IrAndDatalogOperationPairs> {};
+
+// Check that the IrOperation is lowered to the expected `IsOperationFact`.
+// Rather than introducing a `operator==` for this, we just rely on them
+// producing the same `ToDatalogString` result.
+TEST_P(IrOperationLoweringTest, IrOperationLoweringTest) {
+  const auto [ir_operation, datalog_is_operation] = GetParam();
+
+  ir::SsaNames ssa_names;
```

Comment:
Isn't this defined in the visitor?

Reply:
No, the visitor takes an `ssa_names` as a parameter. It might not matter for now, but I guess I was thinking that if we had multiple TUs or language frontends that wished to do lowering, it might make sense to use the same `SsaNames` for all of them. I can make it a member if you think that's not worth the extra hassle at the moment.

Reply:
I see your point, but let us make it a member for now? It would be good to avoid unnecessary global state between calls to visitor.

Reply:
Done.